File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    
    let maxProfit = 0;
    
    let l = 0;
    let r = 1;

    while (r < prices.length) {
        const left = prices[l];
        const right = prices[r];

        if (right < left) {
            l = r;
            r++;
            continue;
        }

        const curr = right - left;
        maxProfit = Math.max(maxProfit, curr);
        r++;
    }

    return maxProfit;
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0

        l = 0
        r = 0

        while r < len(prices):
            if prices[r] > prices[l]: profit = max(profit, (prices[r] - prices[l]))
            else: l = r
            r += 1
        return profit